fvm/executionParameters.go

```diff
@@ -16,43 +18,105 @@ import (
 	"github.com/onflow/flow-go/fvm/utils"
 )
 
-func getEnvironmentMeterParameters(
+type MeterParamOverrides struct {
+	ComputationWeights meter.ExecutionEffortWeights // nil indicates no override
+	MemoryWeights      meter.ExecutionMemoryWeights // nil indicates no override
+	MemoryLimit        *uint64                      // nil indicates no override
+}
+
+func getMeterParameters(
 	ctx Context,
+	proc Procedure,
 	view state.View,
 	derivedTxnData *programs.DerivedTransactionData,
-	params meter.MeterParameters,
 ) (
 	meter.MeterParameters,
 	error,
 ) {
-	sth := state.NewTransactionState(
+	procParams := meter.DefaultParameters().
+		WithComputationLimit(uint(proc.ComputationLimit(ctx))).
+		WithMemoryLimit(proc.MemoryLimit(ctx)).
+		WithEventEmitByteLimit(ctx.EventCollectionByteSizeLimit)
+
+	txnState := state.NewTransactionState(
 		view,
 		state.DefaultParameters().
 			WithMaxKeySizeAllowed(ctx.MaxStateKeySize).
 			WithMaxValueSizeAllowed(ctx.MaxStateValueSize).
 			WithMaxInteractionSizeAllowed(ctx.MaxStateInteractionSize))
 
-	sth.DisableAllLimitEnforcements()
+	// TODO(patrick): cache meter param overrides
+	overrides, err := getMeterParamOverrides(
+		ctx,
+		txnState,
+		derivedTxnData)
+	if err != nil {
+		return procParams, err
+	}
+
+	if overrides.ComputationWeights != nil {
+		procParams = procParams.WithComputationWeights(
+			overrides.ComputationWeights)
+	}
+
+	if overrides.MemoryWeights != nil {
+		procParams = procParams.WithMemoryWeights(overrides.MemoryWeights)
+	}
 
-	env := NewScriptEnv(context.Background(), ctx, sth, derivedTxnData)
+	if overrides.MemoryLimit != nil {
+		procParams = procParams.WithMemoryLimit(*overrides.MemoryLimit)
+	}
 
-	return fillEnvironmentMeterParameters(ctx, env, params)
+	if proc.ShouldDisableMemoryAndInteractionLimits(ctx) {
+		procParams = procParams.WithMemoryLimit(math.MaxUint64)
+	}
+
+	return procParams, nil
 }
 
-func fillEnvironmentMeterParameters(
+func getMeterParamOverrides(
 	ctx Context,
-	env environment.Environment,
-	params meter.MeterParameters,
+	txnState *state.TransactionState,
+	derivedTxnData *programs.DerivedTransactionData,
 ) (
-	meter.MeterParameters,
+	MeterParamOverrides,
 	error,
 ) {
+	var overrides MeterParamOverrides
+	var err error
+	txnState.RunWithAllLimitsDisabled(func() {
+		overrides, err = _getMeterParamOverrides(
+			ctx,
+			txnState,
+			derivedTxnData)
+	})
+
+	if err != nil {
+		return overrides, fmt.Errorf(
+			"error getting environment meter parameter overrides: %w",
+			err)
+	}
 
+	return overrides, nil
+}
+
+func _getMeterParamOverrides(
+	ctx Context,
+	txnState *state.TransactionState,
+	derivedTxnData *programs.DerivedTransactionData,
+) (
+	MeterParamOverrides,
+	error,
+) {
 	// Check that the service account exists because all the settings are
 	// stored in it
 	serviceAddress := ctx.Chain.ServiceAddress()
 	service := runtime.Address(serviceAddress)
 
+	env := NewScriptEnv(context.Background(), ctx, txnState, derivedTxnData)
```

Comment:
For scripts procedures it might make sense to use `ScriptProcedure.RequestContext`.

Reply:
the param parsing overhead is constant so it shouldn't be problematic.